### PR TITLE
Categorization

### PR DESCRIPTION
**Pull request by bauersimon**

### CODE REVIEW

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 1]
```diff
@@ -109,12 +109,14 @@ func (command *Evaluate) Execute(args []string) (err error) {
 
 	// Check that models and languages can be evaluated by executing the "plain" repositories.
 	log.Printf("Checking that models and languages can be used for evaluation")
-	assessmentsPerModel := map[string]metrics.Assessments{}
+	assessmentsPerModelPerRepository := map[string]map[string]metrics.Assessments{}
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:112]: Should we split by "language" as well right away?

Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:112]: Should we introduce identifier types for models, languages, ... ? Would make them easier to read and allows to not mix? Kind ... of ... type ... safety?

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 2]
```diff
@@ -163,7 +165,7 @@ func (command *Evaluate) Execute(args []string) (err error) {
 				language := language.Languages[languageID]
 
 				assessment, ps, err := evaluate.EvaluateRepository(command.ResultPath, model, language, command.TestdataPath, filepath.Join(languageID, repository.Name()))
-				assessmentsPerModel[model.ID()].Add(assessment)
+				assessmentsPerModelPerRepository[model.ID()][repository.Name()] = metrics.Merge(assessmentsPerModelPerRepository[model.ID()][repository.Name()], assessment)
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:168]: Please use and .Add

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:168]: The bigger these maps get, the worst is the copying

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:168]: And it is then easier to read what we do here

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 3]
```diff
@@ -172,13 +174,21 @@ func (command *Evaluate) Execute(args []string) (err error) {
 		}
 	}
 
+	// Combine multiple repository assessments for reporting.
+	assessmentsPerModel := make(map[string]metrics.Assessments, len(assessmentsPerModelPerRepository))
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:178]: Can you print assessmentsPerModelPerRepository before to the CSV and then do the logic for assessmentsPerModel in one block to print the scores.

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:178]: I think that is better grouping: print overall results first and then combine into different reports.

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 2 above)
Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate.go:168]: The assessments per repository are never initialized so the object to `Add` to is `nil`. So I guess I need to manually initialize all of them first?

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:168]: We are doing that for "plain" as well, so i say, yes?

File: evaluate/metrics/assessment.go [hunk 4]
```diff
@@ -137,28 +137,33 @@ func (a Assessments) StringCSV() (row []string) {
 }
 
 func csvHeader() []string {
```
Comment (zimmski) [inline evaluate/metrics/assessment.go:139]: document

**Comment by bauersimon** [comment]
todo: add category unknown if total score is not known

File: evaluate/metrics/assessment.go [hunk 5]
```diff
@@ -137,28 +137,33 @@ func (a Assessments) StringCSV() (row []string) {
...
-	return append([]string{"model", "score"}, allAssessmentKeysStrings...)
+	return append([]string{"model", "repository", "score"}, allAssessmentKeysStrings...)
 }
 
 // FormatStringCSV formats the given assessment metrics as CSV.
```
Comment (zimmski) [inline evaluate/metrics/assessment.go:143]: Should we call this just `FormatCSV` it is the only one. so no need to give it a type

File: evaluate/metrics/assessment_test.go [hunk 6]
```diff
@@ -179,41 +179,47 @@ func TestFormatStringCSV(t *testing.T) {
 	validate(t, &testCase{
 		Name: "Single Empty Model",
 
-		AssessmentPerModel: map[string]Assessments{
-			"Model": Assessments{},
+		AssessmentPerModelPerRepository: map[string]map[string]Assessments{
+			"Model": map[string]Assessments{
+				"Repository": NewAssessments(),
```
Comment (zimmski) [inline evaluate/metrics/assessment_test.go:184]: usually when we do not care we always prefix with "some" -> `some-repository`

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 7]
```diff
@@ -182,8 +182,12 @@ func (command *Evaluate) Execute(args []string) (err error) {
 		}
 	}
 
+	isPlain := len(commandRepositories) == 0 || (len(commandRepositories) == 1 && commandRepositories["plain"])
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:185]: isPlain -> isPlainRepository

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:185]: Wait. what is the logic here? Why do we not ccategorize every repository? I know that we currently only have "plain" Go repository, but why do we already block new ones here?

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:185]: (don't think that this is necessary)

File: evaluate/metrics/category.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	// AssessmentCategoryResponseNoCode indicates that a model's response did not contain any source code.
+	AssessmentCategoryResponseNoCode = AssessmentCategory("response-no-code")
+	// AssessmentCategoryCodeInvalid indicates that a model's generated code produced an error when executed.
+	AssessmentCategoryCodeInvalid = AssessmentCategory("code-invalid")
+	// AssessmentCategoryCodeExecuted indicates that a model's generated code could be executed without an error.
+	AssessmentCategoryCodeExecuted = AssessmentCategory("code-executed")
+	// AssessmentCategoryCodeCoverageStatementReached indicates that a model's generated code reached 100% statement coverage.
+	AssessmentCategoryCodeCoverageStatementReached = AssessmentCategory("code-coverage-statement")
+	// AssessmentCategoryCodeNoExcess indicates that a model's response did not contain more content than requested.
+	AssessmentCategoryCodeNoExcess = AssessmentCategory("code-no-excess")
```
Comment (zimmski) [inline evaluate/metrics/category.go:20]: Perfect category is missing?

Comment (zimmski) [reply evaluate/metrics/category.go:20]: I guess it is `AssessmentCategoryCodeNoExcess`?

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 7 above)
Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate.go:185]: cause for others we don't know the total score... will introduce a category unknown in that case

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 1 above)
Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate.go:112]: Could also use the models and languages directly instead of IDs. That way its type safe already.

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:112]: i mean more in the sense of something like this

someMap[modelID]...

would then fail if the key of the map is of type repositoryID

Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate.go:112]: Could also work. But why introduce extra types if we could just do:

```go
// AssessmentPerModelPerLanguagePerRepository holds a collection of assessments per model per language and per repository.
type AssessmentPerModelPerLanguagePerRepository map[model.Model]map[language.Language]map[string]Assessments
```

Comment (zimmski) [reply cmd/eval-dev-quality/cmd/evaluate.go:112]: Sure!

File: language/testing/mock.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import (
+	"log"
+
+	"github.com/stretchr/testify/mock"
+
+	"github.com/symflower/eval-dev-quality/language"
+)
+
+// MockLanguage is a mocked language.
+type MockLanguage struct {
```
Comment (zimmski) [inline language/testing/mock.go:12]: Holy wow. that was a big commit!

**Review by zimmski — APPROVED** [review]

**Pull request merged**